django/cantusdb_project/main_app/views/sequence.py

```diff
@@ -62,3 +65,27 @@ def form_valid(self, form):
             "Sequence updated successfully!",
         )
         return super().form_valid(form)
+
+    def test_func(self):
+        user = self.request.user
+        sequence_id = self.kwargs.get(self.pk_url_kwarg)
+        sequence = Sequence.objects.get(id=sequence_id)
+        # find the source of this sequence
+        source = Source.objects.get(sequence=sequence)
```

Comment:
Yes it works but, I guess that `source=sequence.source` would be cleaner (I forgot that sequence had a source field...). I don't see any problems with `source=sequence.source` because you're just assigning `sequence.source`, a `Source` object, to a variable called `source`

Reply:
Ah, sorry. I understand what you were trying to say. Yes, this is a clear oversight from my part, but I believe it was working when I was testing it out with various accounts with different privileges. In hindsight, I don't know why it worked or maybe I tested it out on too small of a sample and it happened to work for those. I'll make sure to fix it with your suggestion.

Reply:
If it didn't throw an error, perhaps the `test_func` is not called by anything. Please test more to make sure it does work as expected.

Reply:
**Update**: As talked about with Junhao, Django does not seem to throw an error when we erroneously called `source = Source.objects.get(sequence=sequence)`, even though sequence is not defined as a field in the Source model. By printing out the object assigned to `source`, I was able to see that source was actually correctly assigned the Source object that I was expecting. Perhaps this may be a Django `.get()` method implementation that I am unaware of. However, I have changed `Source.objects.get(sequence=sequence)` and `Source.objects.get(chant=chant)` to `sequence.source` and `chant.source` respectively, for clarity.